mava/systems/tf/variable_utils.py

```diff
@@ -0,0 +1,175 @@
+        self._add_future: Optional[futures.Future] = None
+
+    def get_async(self) -> None:
+        """Asynchronously updates the get variables with the latest copy from source."""
+
+        # Track the number of calls (we only update periodically).
+        if self._get_call_counter < self._get_update_period:
+            self._get_call_counter += 1
+
+        period_reached: bool = self._get_call_counter >= self._get_update_period
+
+        if period_reached and self._get_future is None:
+            # The update period has been reached and no request has been sent yet, so
+            # making an asynchronous request now.
+            self._get_future = self._async_request()  # type: ignore
+            self._get_call_counter = 0
+
+        if self._get_future is not None and self._get_future.done():
+            # The active request is done so copy the result and remove the future.\
+            self._copy(self._get_future.result())
+            self._get_future = None
+
+        return
+
+    def set_async(self) -> None:
+        """Asynchronously updates source with the set variables."""
+        # Track the number of calls (we only update periodically).
+        if self._set_call_counter < self._set_update_period:
+            self._set_call_counter += 1
+
+        period_reached: bool = self._set_call_counter >= self._set_update_period
+
+        if period_reached and self.set_future is None:  # type: ignore
+            # The update period has been reached and no request has been sent yet, so
+            # making an asynchronous request now.
+            self.set_future = self._async_adjust()  # type: ignore
+            self._set_call_counter = 0
+            return
+        if self.set_future is not None and self.set_future.done():
+            self.set_future = None  # type: ignore
+        return
+
+    def add_async(self, names: List[str], vars: Dict[str, Any]) -> None:
+        """Asynchronously adds to source variables."""
+        if self._add_future is not None and self._add_future.done():
+            self._add_future = None
+
+        if self._add_future is None:
+            # The update period has been reached and no request has been sent yet, so
+            # making an asynchronous request now.
+            self._add_future = self._async_add(names, vars)  # type: ignore
+            return
+        else:
+            tf.print("We got a problem!")
+            exit()
+
+        return
+
+    def add_and_wait(self, names: List[str], vars: Dict[str, Any]) -> None:
+        """Adds the specified variables to the corresponding variables in source
+        and waits for the process to complete before continuing."""
+        self._client.add_to_variables(names, vars)
+
+    def get_and_wait(self) -> None:
+        """Updates the get variables with the latest copy from source
+        and waits for the process to complete before continuing."""
+        self._copy(self._request())  # type: ignore
+        return
```

Comment:
I asked this in another comment, does `return` have any effect or is this the same as acme's `update_and_wait` (https://github.com/deepmind/acme/blob/e4b6046f7ab70b2cf1769847605f06059800f599/acme/tf/variable_utils.py#L97)?

Reply:
Please see the above comment. The return is just to make mypy happy. It has no effect.